compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/CommitmentDispatchCodeGenerator.kt

```diff
@@ -0,0 +1,78 @@
+package edu.cornell.cs.apl.viaduct.codegeneration
+
+import com.squareup.kotlinpoet.CodeBlock
+import edu.cornell.cs.apl.viaduct.backends.commitment.Commitment
+import edu.cornell.cs.apl.viaduct.errors.CodeGenerationError
+import edu.cornell.cs.apl.viaduct.selection.ProtocolCommunication
+import edu.cornell.cs.apl.viaduct.syntax.Host
+import edu.cornell.cs.apl.viaduct.syntax.Protocol
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.AtomicExpressionNode
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.LetNode
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.SimpleStatementNode
+
+class CommitmentDispatchCodeGenerator(
+    val context: CodeGeneratorContext
+) : CodeGenerator {
+    private val commitmentCreatorGenerator = CommitmentCreatorGenerator(context)
+    private val commitmentHolderGenerator = CommitmentHolderGenerator(context)
+
+    override fun guard(protocol: Protocol, expr: AtomicExpressionNode): CodeBlock =
+        when (protocol) {
```

Comment:
Pull this logic into a function. See `CodeGenerator.unions` for an example.

Reply:
Is this class necessary? You can do the dispatching when you are populating the code generator map. For example,

```
  if (protocol is Commitment)
      initGeneratorMap[protocol] = CommitmentDispatchCodeGenerator(context)
```

can be

```
  if (protocol is Commitment) {
      initGeneratorMap[protocol] =
      when {
           host == protocol.cleartextHost -> CommitmentCreatorCodeGenerator(context)
           protocol.hashHosts.contains(host) -> CommitmentCreatorCodeGenerator(context)
      }
  }

Reply:
Protocol map logic should not special case on backends. The case analysis that is there will go away soon. So we need some method of dispatching.